src/Input/AbstractInputParam.php

```diff
@@ -0,0 +1,203 @@
+    private $allowedModes = [
+        InputOption::VALUE_NONE,
+        InputOption::VALUE_REQUIRED,
+        InputOption::VALUE_REQUIRED | InputOption::VALUE_IS_ARRAY,
+    ];
+
+    /** @var mixed */
+    private $default;
+
+    /** @var string */
+    private $description = '';
+
+    /**
+     * Parameter name; must be set by class composing trait!
+     *
+     * @var string
+     */
+    private $name;
+
+    /**
+     * InputOption mode to use with this parameter.
+     *
+     * @var int
+     */
+    private $optionMode = InputOption::VALUE_REQUIRED;
+
+    /** @var bool */
+    private $required = false;
+
+    /** @var null|string|array */
+    private $shortcut;
+
+    public function __construct(string $name)
+    {
+        $this->name = $name;
+    }
+
+    /**
+     * Default value to use if none provided.
+     *
+     * @return mixed
+     */
+    public function getDefault()
+    {
+        return $this->default;
+    }
+
+    public function getDescription(): string
+    {
+        return $this->description;
+    }
+
+    public function getName(): string
+    {
+        return $this->name;
+    }
+
+    public function getOptionMode(): int
+    {
+        return $this->optionMode;
+    }
+
+    // phpcs:disable WebimpressCodingStandard.Functions.ReturnType.ReturnValue
+    public function getShortcut()
+    {
+        return $this->shortcut;
+    }
+
+    public function isRequired(): bool
+    {
+        return $this->required;
+    }
+
+    /**
+     * @param mixed $defaultValue
+     */
+    public function setDefault($defaultValue): InputParamInterface
+    {
+        $this->default = $defaultValue;
+        return $this;
+    }
+
+    public function setDescription(string $description): InputParamInterface
+    {
+        $this->description = $description;
+        return $this;
+    }
+
+    public function setOptionMode(int $mode): InputParamInterface
+    {
+        if (! in_array($mode, $this->allowedModes, true)) {
+            throw new InvalidArgumentException(sprintf(
+                'Invalid option mode; must be one of %s::VALUE_NONE,'
+                . ' VALUE_REQUIRED, or VALUE_REQUIRED | VALUE_IS_ARRAY',
+                InputOption::class
+            ));
+        }
+        $this->optionMode = $mode;
+        return $this;
+    }
+
+    // phpcs:disable WebimpressCodingStandard.Functions.Param.MissingSpecification
+    public function setShortcut($shortcut): InputParamInterface
+    {
+        $this->validateShortcut($shortcut);
+        $this->shortcut = $shortcut;
+        return $this;
+    }
+
+    public function setRequiredFlag(bool $required): InputParamInterface
+    {
+        $this->required = $required;
+        return $this;
+    }
+
+    /**
+     * @param mixed $shortcut
+     * @throws InvalidArgumentException When shortcut is an invalid type.
+     * @throws InvalidArgumentException When shortcut is empty.
+     */
+    private function validateShortcut($shortcut): void
```

Comment:
I really do not like symfony way here, as it looks like `['b', 'a|r']` is a valid shortcut, and it will create three shortcuts - `b`, `a`, `r`.

Reply:
Right. I specifically didn't bring all of that over with the intention of being able to simply provide a value verbatim to the `InputOption` that _should_ be valid. With the validation I've written, I'm ensuring that when we create our `InputOption`, we have data it will consider valid, so that the invalid shortcut is identified when we try to set it on the parameter instead. I do that by:

- Allowing `null`.
- Disallowing non-string, non-array values if a value is provided.
- Checking for empty string values (I've updated the check to do a `trim()` operation first that includes both a dash and a space in the character mask).
- Checking for empty array values.
- Checking each item of an array to verify it is a non-empty string.

If it passes that validation, `InputOption` will not raise an exception when we create an instance, which is the goal; how it serializes internally is then not our problem.